src/java/com/puppetlabs/http/client/impl/TimedFutureCallback.java

```diff
@@ -0,0 +1,46 @@
+package com.puppetlabs.http.client.impl;
+
+import com.codahale.metrics.Timer;
+import org.apache.http.concurrent.FutureCallback;
+
+import java.util.ArrayList;
+
+public final class TimedFutureCallback<T> implements FutureCallback<T> {
+
+    private final FutureCallback<T> delegate;
+
+    private final ArrayList<Timer.Context> timerContexts;
+
+    public TimedFutureCallback(FutureCallback<T> delegate, ArrayList<Timer.Context> timerContexts) {
+        this.delegate = delegate;
+        this.timerContexts = timerContexts;
+    }
+
+    public void completed(T result) {
+        if (timerContexts != null) {
```

Comment:
Could put these if blocks in a common helper method that each of the callbacks invoke.